ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Envelope.scala

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils
+
+import java.util.zip.{GZIPInputStream, GZIPOutputStream}
+import com.google.protobuf.ByteString
+import scala.util.Try
+import com.daml.ledger.participant.state.kvutils.{DamlKvutils => Proto}
+
+/** Envelope is a wrapping for "top-level" kvutils messages that provides
+  * versioning and compression and should be used when storing or transmitting
+  * kvutils messages.
+  */
+object Envelope {
+
+  sealed trait Message extends Product with Serializable
+
+  final case class SubmissionMessage(submission: Proto.DamlSubmission) extends Message
+
+  final case class LogEntryMessage(logEntry: Proto.DamlLogEntry) extends Message
+
+  final case class StateValueMessage(value: Proto.DamlStateValue) extends Message
+
+  val kvutilsVersion: Long = 1
+
+  private def newBuilderFor(kind: Proto.Envelope.MessageKind): Proto.Envelope.Builder =
```

Comment:
Don't we need a builder for uncompressed envelopes?

Reply:
Sure, if we want to produce them. I guess I might as well add the uncompresswd variants for symmetry. Might be useful when one knows it doesn’t make sense to compress.